### ISSUE TITLE
Bug in how pb handles synchronous pb.Errors in remote_foo methods

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2317]
|<img alt="bgranger's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| bgranger reported|
|-|-|
|Trac ID|trac#2317|
|Type|defect|
|Created|2006-12-21 00:00:03Z|

According to the twisted docs, any exception that inherits from pb.Error will be passed back across the wire.  However, there are cases where this is not true:

```
 def remote_foo(self, f):
       d = self.service.foo(f)
       def raiser():
           raise pb.Error('I pbfailed')
       d.addErrback(lambda reason: raiser()) # reason = Exception(), not pb.Error
       return d
```

This does not result in the pb.Error being handled in the expected way (this process will complain.  This type of usage case is useful for converting general exceptions to pb.Errors.  

According to Jean-Paul:

"This is a bug in PB's handling of Deferreds which fail with an Error.
They are not handled in the same way as Errors which are raised
synchronously by the method to which the PB message is dispatched."

<details><summary>Searchable metadata</summary>

```
trac-id__2317 2317
type__defect defect
reporter__bgranger bgranger
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__pb__error__failure pb, error, failure
time__1166659203000000 1166659203000000
changetime__1166986680000000 1166986680000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun removed owner|
|-|-|

Fixed this in synchronous-error-2317

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_2"></a>|@jerub commented|
|-|-|

oh wow yay whoop.

there was toilet humour in our unit tests and you eliminated it. thankyou so much. :)

testCopiedFailureLogging needs renaming. I suggest doing that in this branch, otherwise it's the only styleguide violation in test_pbfailure.

There's a huge number of imports here that aren't used. But I think they're because of the external api that pb provides. Here's one that should be removed.
```
twisted/spread/pb.py:71: 'warnings' imported but unused
```

Merry christmas. Please remove the spurious import and merge this branch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set status to `closed`|
|-|-|

(In [19068]) Merge synchronous-error-2317

Author: exarkun
Reviewer: jerub
Fixes [#2317](https://github.com/twisted/twisted/issues/2317)

Fix error handling around remote_* methods so that when they return a Deferred which
eventually errbacks with a pb.Error or pb.Error subclass, the same rules are applied
as are applied to the case where a pb.Error or pb.Error subclass is raised from the
method synchronously (ie, do not log it on the server side before sending it).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
